go_faas_test.go

```diff
@@ -0,0 +1,901 @@
+
+import (
+	"github.com/stretchr/testify/assert"
+	"os"
+	"testing"
+	"time"
+)
+
+func TestCreateSystemFunctions(t *testing.T) {
+	cli, err := NewClient(&FaasGatewayCredentials{
+		Username:       os.Getenv("OPENFAAS_USERNAME"),
+		Password:       os.Getenv("OPENFAAS_PASSWORD"),
+		GatewayAddress: os.Getenv("OPENFAAS_GATEWAY_ADDR"),
+		ClusterType:    os.Getenv("OPENFAAS_CLUSTER_TYPE"),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testcases := []struct {
+		Name          string
+		FunctionDef   *FunctionDefintion
+		StatusCode    int
+		ErrorExpected bool
+	}{
+		{
+			Name: "Function/nodeinfo123456",
+			FunctionDef: &FunctionDefintion{
+				Service:    "nodeinfo123456",
+				Network:    "func_functions",
+				Image:      "functions/nodeinfo:latest",
+				EnvProcess: "node main.js",
+				EnvVars: EnvVars{
+					AdditionalProp1: "string",
+					AdditionalProp2: "string",
+					AdditionalProp3: "string",
+				},
+				Constraints: []string{
+					"node.platform.os == linux",
+				},
+				Labels: map[string]string{
+					"example": "func1",
+				},
+				Annotations: Annotations{
+					Topics: "awesome-kafka-topic",
+					Foo:    "some",
+				},
+				RegistryAuth: "dXNlcjpwYXNzd29yZA==",
+				Limits: Limits{
+					Memory: "128M",
+					CPU:    "0.01",
+				},
+				Requests: Requests{
+					Memory: "128M",
+					CPU:    "0.01",
+				},
+				ReadOnlyRootFilesystem: true,
+			},
+			StatusCode:    202,
+			ErrorExpected: false,
+		},
+		{
+			Name: "Function/yetanothernodeinfo",
+			FunctionDef: &FunctionDefintion{
+				Service:    "yetanothernodeinfo",
+				Network:    "func_functions",
+				Image:      "functions/nodeinfo:latest",
+				EnvProcess: "node main.js",
+				Constraints: []string{
+					"node.platform.os == linux",
+				},
+				Labels: map[string]string{
+					"label": "val",
+				},
+				RegistryAuth: "dXNlcjpwYXNzd29yZA==",
+				Limits: Limits{
+					Memory: "128M",
+					CPU:    "0.01",
+				},
+				Requests: Requests{
+					Memory: "128M",
+					CPU:    "0.01",
+				},
+			},
+			StatusCode:    202,
+			ErrorExpected: false,
+		},
+		{
+			Name: "Function/InvalidService",
+			FunctionDef: &FunctionDefintion{
+				Service:      "InvalidService",
+				Network:      "func_functions",
+				Image:        "",
+				EnvProcess:   "node main.js",
+				RegistryAuth: "dXNlcjpwYXNzd29yZA==",
+				Limits: Limits{
+					Memory: "128M",
+					CPU:    "0.01",
+				},
+				Requests: Requests{
+					Memory: "128M",
+					CPU:    "0.01",
+				},
+			},
+			StatusCode:    400,
+			ErrorExpected: true,
+		},
+	}
+
+	for _, c := range testcases {
+		t.Run(c.Name, func(t *testing.T) {
+			resp, err := cli.CreateSystemFunctions(c.FunctionDef)
+			if err != nil {
+				t.Errorf("error while calling %s/system/functions: %v", os.Getenv("OPENFAAS_GATEWAY_ADDR"), err)
+			}
+			assert.Equal(t, nil, err)
+			assert.Equal(t, c.StatusCode, resp.StatusCode)
+		})
+	}
+}
+
+func TestGetSystemFunctions(t *testing.T) {
+	cli, err := NewClient(&FaasGatewayCredentials{
```

Comment:
May be its better if we create a TestSuite and initialize all the functions to use the suite definition. https://godoc.org/github.com/stretchr/testify/suite

If it is hard, you can simply write a SetupTest method and inject that for every test

Reply:
Oh yes! I was looking for exactly the same thing. I tried to do this on my own but `testify/suite` seems so much better

Reply:
Also, we need to see the 2 functions I mentioned above which are not performing as expected